textscreen/txt_sdl.c

```diff
@@ -0,0 +1,1009 @@
+        font = &highdpi_font;
+    }
+}
+
+//
+// Initialize text mode screen
+//
+// Returns 1 if successful, 0 if an error occurred
+//
+
+void TXT_PreInit(SDL_Window *preset_window, SDL_Renderer *preset_renderer)
+{
+    if (preset_window != NULL && preset_renderer != NULL)
+    {
+        TXT_SDLWindow = preset_window;
+        renderer = preset_renderer;
+    }
+}
+
+int TXT_Init(void)
+{
+    int flags = 0;
+
+    if (SDL_Init(SDL_INIT_VIDEO) < 0)
+    {
+        return 0;
+    }
+
+    ChooseFont();
+
+    screen_image_w = TXT_SCREEN_W * font->w;
+    screen_image_h = TXT_SCREEN_H * font->h;
+
+    // If highdpi_font is selected, try to initialize high dpi rendering.
+    if (font == &highdpi_font)
+    {
+        flags |= SDL_WINDOW_ALLOW_HIGHDPI;
+    }
+
+    if (TXT_SDLWindow == NULL)
+    {
+    TXT_SDLWindow =
+        SDL_CreateWindow("", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED,
+                         screen_image_w, screen_image_h, flags);
+    }
+
+    if (TXT_SDLWindow == NULL)
+        return 0;
+
+    // Destroy the existing renderer, so we can create our own new one
+
+    if (renderer != NULL)
+    {
+        SDL_DestroyRenderer(renderer);
+    }
+
+    renderer = SDL_CreateRenderer(TXT_SDLWindow, -1, SDL_RENDERER_PRESENTVSYNC);
+
+    if (renderer == NULL)
+        renderer = SDL_CreateRenderer(TXT_SDLWindow, -1, SDL_RENDERER_SOFTWARE);
+
+    if (renderer == NULL)
+        return 0;
+
+    // Special handling for OS X retina display. If we successfully set the
+    // highdpi flag, check the output size for the screen renderer. If we get
+    // the 2x doubled size we expect from a retina display, use the large font
+    // for drawing the screen.
+    if ((SDL_GetWindowFlags(TXT_SDLWindow) & SDL_WINDOW_ALLOW_HIGHDPI) != 0)
+    {
+        int render_w, render_h;
+
+        if (SDL_GetRendererOutputSize(renderer, &render_w, &render_h) == 0
+         && render_w >= TXT_SCREEN_W * large_font.w
+         && render_h >= TXT_SCREEN_H * large_font.h)
+        {
+            font = &large_font;
+            // Note that we deliberately do not update screen_image_{w,h}
+            // since these are the dimensions of textscreen image in screen
+            // coordinates, not pixels.
+        }
+    }
+
+    // Failed to initialize for high dpi (retina display) rendering? If so
+    // then use the normal resolution font instead.
+    if (font == &highdpi_font)
+    {
+        font = &normal_font;
+    }
+
+    // Instead, we draw everything into an intermediate 8-bit surface
+    // the same dimensions as the screen. SDL then takes care of all the
+    // 8->32 bit (or whatever depth) color conversions for us.
+    screenbuffer = SDL_CreateRGBSurface(0,
+                                        TXT_SCREEN_W * font->w,
+                                        TXT_SCREEN_H * font->h,
+                                        8, 0, 0, 0, 0);
+
+    SDL_RenderSetLogicalSize(renderer, screenbuffer->w, screenbuffer->h);
```

Comment:
Comment, please! If there are other parts of this library that you had to change to achieve the current state, please comment them as well.